### ISSUE TITLE
Infer missing nominal attribute from start value

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8084]
### Description

It is well-known that providing appropriate scaling values improves the robustness of iterative nonlinear solvers, see [paper](http://dx.doi.org/10.1145/3158191.3158192). Well-designed libraries should address this issue by providing appropriate nominal attributes wherever needed. However, this is sometimes difficult, inconvenient, or simply overlooked.

Suppose that, after nominal attribute propagation through equations has been carried out (see [#4515](https://trac.openmodelica.org/OpenModelica/ticket/4515)), there are some variables that are (possibly) going to be selected as iteration variables and only have a non-zero start attribute, but not a nominal attribute, e.g.
```
Real h(start = 500000);
```
In this case, the prior information from the model developer is obviously that `h` will have the order of magnitude of half a million, so 500000 will be a much better nominal attribute (for scaling) than the currently used value of one.

This could be implemented in the runtime, using the start attribute to set the scaling value, when there is none. But maybe it would be better implemented in the backend, which could also propagate it to other similar variables based on connection equations.

**Comment by casella** [comment]
@phannebohm any chance you can take care of this anytime soon, so we get it in 1.19.0?

Thanks

**Comment by phannebohm** [comment]
I don't think so. I'll be on vacation now. This needs some thought, I think. As you already say in #4515, some non-trivial research might await. If we can get some MWEs I might do simple things starting next Year. But the trivial stuff seems to be done already.

I could imagine all kinds of weird expressions that make it unclear how to propagate the nominal value, e.g. some polynomial relationship `y = 1e5*(x-10)^2 + 1e3` and `x.nominal=10`. Just plugging in `x=10` would result in `y.nominal = 1e3` but slight deviations in `x` show that `y.nominal` should actually be `1e5` or something in between... maybe this is an unrealistic scenario, so some practical examples would help. But I think it is clear that we cannot blindly compute some expressions.

And while we are at it, I would like to also propagate min and max attributes, those might perhaps be used to ensure that some variables are never zero for example, or other things. I guess those topics are related: maybe having a range of values lets us estimate nominal values more confidently.

The ideas seem easy but I suspect some traps are around the corner.

**Comment by casella** [comment]
> I don't think so. I'll be on vacation now.

Sorry, I forgot about that 😃 . Don't worry, this is not urgent.

> I could imagine all kinds of weird expressions that make it unclear how to propagate the nominal value, e.g. some polynomial relationship `y = 1e5*(x-10)^2 + 1e3` and `x.nominal=10`. Just plugging in `x=10` would result in `y.nominal = 1e3` but slight deviations in `x` show that `y.nominal` should actually be `1e5` or something in between... maybe this is an unrealistic scenario, so some practical examples would help. But I think it is clear that we cannot blindly compute some expressions.

I guess you misunderstood me. In this ticket I'm not asking to propagate nominal values through expressions, that would just be hopeless. With reference to the case you mentioned, if `y.start = 1e4` and `y.nominal` is not set, I would set `y.nominal = y.start`, i.e. 1e4. The only exception would be if `y.start = 0`, because then it is obvious that 0 is not a good choice for `y.nominal`.

This should be quite straightforward.

**Comment by phannebohm** [comment]
Mi scusi, of course, it even says so in the title. :sweat_smile:
That seems reasonable, and possibly quick to do. Let's try it next month.

**Comment by casella** [comment]
😄
